### DISCUSSION TITLE
Use of cert manager instead of generating CA bundle during build

### DISCUSSION THREAD

**Discussion by anekdoti** [dotnet/dotnet-operator-sdk#319] [Q&A]
The [documentation for webhooks](https://buehler.github.io/dotnet-operator-sdk/docs/webhooks.html) advises to just add the generated files `ca.pem` and `ca-key.pem` to the repository of your operator, and (implicitly) to add them to your GitOps repository together with the kustomize script that generates a secret out of them. I understand that it is no security in this case, as the CA is only used cluster-internally, but it is still somehow an anti pattern.

In our cluster infrastructure, we have the [cert manager](https://cert-manager.io/docs/) available. Do you have any hints how to generate the CA + key using cert manager using the custom resource "Certificate" of cert manager? Did somebody already try this and maybe even have a sample resource? What are things to keep an eye on?

**Comment by buehler** [comment]
Hmm. I see the point. I just learned that Kubernetes has its own PKI, but that does not automatically approve CSRs.

So theoretically you could overwrite the secret that is generated with the CA cert/key combination. I don't know if it is easily possible, but you could try to patch the deployment resource in a way to overwrite the required secret that mounts the certificate. That way, you can generate the certificate with the cert manager and then mount it into the operator.

But this is a good idea that we create a configuration for this. I assume there are others that would use "real" certificates.

**Comment by erin-allison** [comment]
That seems like a documentation error to me. If you use the auto-generated operator definition, there is an init container which pulls down cfssl and generates a set of certificates for the webhook to use.

https://github.com/buehler/dotnet-operator-sdk/blob/02bb7ecd4309cd26ec4b3cdd60b40584fd633b94/src/KubeOps/Operator/Commands/Generators/OperatorGenerator.cs#L119-L153

https://github.com/buehler/dotnet-operator-sdk/blob/02bb7ecd4309cd26ec4b3cdd60b40584fd633b94/src/KubeOps/Operator/Commands/Management/Webhooks/Install.cs#L54-L72

https://github.com/buehler/dotnet-operator-sdk/blob/02bb7ecd4309cd26ec4b3cdd60b40584fd633b94/src/KubeOps/Operator/Commands/CommandHelpers/CertificateGenerator.cs#L93-L128

**Reply by erin-allison** [reply]
You could use cert-manager if you wanted to customize the deployment files; ~~you would simply need to remove the init container and mount a secret as a folder (or files) in the right spot on the operator pod(s).~~

EDIT: I realized after a bit more reading that actually wouldn't work as well. The init container does more than just certificates; it is responsible for creating and applying the webhook definitions.

**Reply by erin-allison** [reply]
@buehler I do agree with allowing certificate generation to be disabled, but would like to see that accompanied with details of where exactly the external certificate files would need to be mounted (and what they would need to be named) to have the webhook still function.

**Reply by buehler** [reply]
That is a fine idea. If we just disable the generation of the certs, a user can create and mount them on their own. Let me think about this.

**Reply by erin-allison** [reply]
The best way to do that would probably be an additional argument for the `webhook install` function, as the decision on whether or not to generate certificates is more at run-time, not design-time. From there, someone could just tack the argument on in the generated `deployment.yaml`
